### PR TITLE
Add Rust to SCons

### PR DESCRIPTION
**Pull request by PeanutbutterWarrior**
Add Rust to the SCons build system. If a _Cargo.toml_ file is present then it builds with Cargo, otherwise it will build with rustc. As such a _Cargo.toml_ file is necessary if any external crates are necessary (usally an `extern crate ...;` in the source file). This could be replaced with a custom `Scanner`, but that's probably overkill and error prone.

In order to have all the code files get picked up by SCons, I have the _*.rs_ and the _Cargo.toml_ file in the same directory, and define the path to the source file in _Cargo.toml_. An example of this is the Cooley-Tukey code.

I have the _*.pdb_ files and the _target/_ directories marked by SCons for cleaning when using the `-c` argument. This deletes the _Cargo.lock_ files, causing cargo to recompile the external crates on the next build.

Due to Cargo putting the executable into a directory in the target directory, it is built in the same place as the source file and the executable then copied to the _build/_ directory. There is a command line argument to copy the build artifacts to a directory (--out-dir) but this is only available on nightly rust. When (and if) this gets stabilized then it would simplify the SConstruct file to take advantage of this.

### CODE REVIEW

**Comment by PeanutbutterWarrior** [comment]
Cooley-Tukey and Split Operator Method don't compile, but this is a problem with those files rather than SCons. I've already 
opened an issue for it.

File: contents/monte_carlo_integration/code/rust/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rand = "0.8.4"
+
+[[bin]]
+path = "./monte_carlo.rs"
+name = "main"
```
Comment (ShadowMitia) [inline contents/monte_carlo_integration/code/rust/Cargo.toml:13]: I would use the file name to make things clearer.

```suggestion
name = "monte_carlo"
```

File: contents/split-operator_method/code/rust/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+num = "0.4.0"
+rustfft = "5.1.1"
+
+[[bin]]
+path = "./split_op.rs"
+name = "main"
```
Comment (ShadowMitia) [inline contents/split-operator_method/code/rust/Cargo.toml:14]:
```suggestion
name = "split_op"
```

File: contents/huffman_encoding/code/rust/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+itertools = "0.10.1"
+
+[[bin]]
+path = "./huffman.rs"
+name = "main"
```
Comment (ShadowMitia) [inline contents/huffman_encoding/code/rust/Cargo.toml:13]:
```suggestion
name = "huffman"
```

File: contents/cooley_tukey/code/rust/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+rand = "0.8.4"
+rustfft = "6.0.1"
+
+[[bin]]
+path = "./fft.rs"
+name = "main"
```
Comment (ShadowMitia) [inline contents/cooley_tukey/code/rust/Cargo.toml:14]:
```suggestion
name = "fft"
```

File: SConstruct [hunk 5]
```diff
@@ -10,13 +10,22 @@ To run the compilation for all implementations in one language, e.g. C, run the
...
+
+rust_rustc_builder = Builder(action='rustc $SOURCE -o $TARGET$PROGSUFFIX')
+
+env = Environment(ENV={'PATH': os.environ['PATH']},
+                  BUILDERS={'rustc': rust_rustc_builder, 'cargo': rust_cargo_builder},
+                  tools=['gcc', 'gnulink', 'g++', 'gas'])
 
 env['CCFLAGS'] = ''
 env['CXXFLAGS'] = '-std=c++17'
+env['ASFLAGS'] = '--64'
```
Comment (ShadowMitia) [inline SConstruct:28]: I don't know `scons` why is this needed for Rust?

File: contents/barnsley/code/rust/Cargo.toml [hunk 6]
```diff
@@ -1,9 +1,13 @@
...
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-rand = "0.8.4"
\ No newline at end of file
+rand = "0.8.4"
+
+[[bin]]
+path = "./barnsley.rs"
+name = "main"
```
Comment (ShadowMitia) [inline contents/barnsley/code/rust/Cargo.toml:13]: I would use the file name to make things clearer.

```suggestion
name = "barnsley"
```

**Review by ShadowMitia — COMMENTED** [review]
I wonder if we shouldn't keep the `src/` to not diverge too much from default `Cargo.toml` settings. We still want to show good practices with Rust.

I see there are builders for both `rustc` and `cargo`, which is fine. But maybe it'll be simplest if we just keep the `cargo` one?

File: SConstruct (hunk 5 above)
Comment (PeanutbutterWarrior) [reply SConstruct:28]: That's part of the assembly build. It might have got mixed up in a rebase

File: contents/monte_carlo_integration/code/rust/Cargo.toml (hunk 1 above)
Comment (PeanutbutterWarrior) [reply contents/monte_carlo_integration/code/rust/Cargo.toml:13]: The name determines what the name of the executable is. If it's not constant then it will need to be passed as another arguement to the build, and as it gets renamed automatically when it's moved I thought this would be simpler.

**Comment by ShadowMitia** [comment]
Dockerfile install looks fine to me.

**Review by Amaras — APPROVED** [review]
Alright, I tested the Rust compilation with your docker file, and it works (except the two you pointed out in #944, of course).

I'm happy to merge this in when you are ready.﻿

**Comment by PeanutbutterWarrior** [comment]
I've fixed the issues with some programs not compiling due to the wrong library versions being used. Solves #944 . It could do with updating, but that's another issue.

**Review by Amaras — APPROVED** [review]
Can confirm it works in the docker container.

Happy to merge this :)﻿
